crates/core/src/node/op_state_manager.rs

```diff
@@ -130,14 +169,92 @@ impl OpManager {
             ring,
             ops,
             to_event_listener: notification_channel,
-            ch_outbound,
+            ch_outbound: Arc::new(ch_outbound),
             new_transactions,
             result_router_tx,
             peer_ready,
             is_gateway,
+            sub_operations,
+            root_ops_awaiting_sub_ops: Arc::new(DashMap::new()),
+            parent_of,
+            expected_sub_operations,
+            failed_parents,
         })
     }
 
+    /// Marks a child operation as completed and decrements the expected counter.
+    /// Removes tracking entry when all expected children have completed.
+    fn mark_sub_op_completed(&self, root_tx: Transaction) {
+        let mut should_remove = false;
+        if let Some(mut expected) = self.expected_sub_operations.get_mut(&root_tx) {
+            if *expected > 0 {
+                *expected -= 1;
+                tracing::debug!(
+                    root_tx = %root_tx,
+                    remaining = *expected,
+                    "sub-operation completed"
+                );
+            }
+            if *expected == 0 {
+                should_remove = true;
+            }
+        }
+
+        if should_remove {
+            self.expected_sub_operations.remove(&root_tx);
+        }
+    }
+
+    fn remove_child_link(&self, parent: Transaction, child: Transaction) {
+        let should_remove_parent = if let Some(mut children) = self.sub_operations.get_mut(&parent)
+        {
+            children.remove(&child);
+            let is_empty = children.is_empty();
+            drop(children);
+            is_empty
+        } else {
+            false
+        };
+
+        if should_remove_parent {
+            self.sub_operations.remove(&parent);
+        }
+
+        self.parent_of.remove(&child);
+    }
+
+    fn cleanup_parent_tracking(&self, parent: Transaction) {
+        self.expected_sub_operations.remove(&parent);
+        self.sub_operations.remove(&parent);
+    }
+
+    fn spawn_client_result(&self, tx: Transaction, host_result: HostResult) {
+        let router_tx = self.result_router_tx.clone();
+        let notifier = self.to_event_listener.clone();
+        GlobalExecutor::spawn(async move {
+            if let Err(err) = router_tx.send((tx, host_result)).await {
+                tracing::error!(
```

Comment:
I think can be warn too to reduce the noise; anyway if this fails there will be more errors and the node will be likely shutting down